Archive/Draft/RFCxxxx-PS7-Bootstrapper.md

```diff
@@ -0,0 +1,256 @@
+### .ps1 registration
+
+`pwsh.exe` will be registered to execute .ps1 scripts, by default.
+However, it will have code to detect if started from the `explorer.exe` and instead
+give an error message that scripts need to be run from the console and not from Explorer.
+This means that from cmd.exe, you can execute .ps1 scripts without explicitly calling `pwsh.exe` and that applications
+can start depending on .ps1 scripts instead of wrapping them in batch files.
+The default execution policy on both Client and Server SKUs should be RemoteSigned.
+
+### Native code
+
+We can't rely on .NET Framework being available on some Windows skus we want to support and keeping the bootstrapper as
+small as possible, we don't want to rely on .NET (Core) and ship that runtime.
+Also, we want the bootstrapper to have minimal dependencies as future Windows SKUs may not ship with certain dependencies.
+
+#### Rust
+
+Rust has the benefit over C/C++ of having well defined contracts for resource management thus not having issues that persist in C/C++ code
+with regards to resource management which results in not only leaks but security issues.
+Resulting Rust binary code is very similar in size and performance as C/C++.
+Rust built code is supported in Windows.
+
+Rust uses the MS C runtime at execution, but we should statically link so that there aren't external dependencies that may be missing on the system.
+
+Need to evaluate what APIs are available on the smallest Windows SKU for downloading.
+
+This is the current preferred option.
+
+### Installer
+
+MSI has dependencies that are not available on Windows SKUs so would not be appropriate as the installer for the bootstrapper.
+Instead, we should look at using [Squirrel](https://github.com/Squirrel/Squirrel.Windows) which supports installing even if the product (existing version)
+is already in use.
+
+The bootstrapper would download the installer using Windows Update as the store to avoid issues with enterprises not allowing
+network access to other sites.
+The installer and binaries may need to be Windows signed instead of Microsoft signed.
+Packages of PS7 outside of this installer used by the boostrapper would continue to be Microsoft signed.
+
+### User Experience
+
+Since the initial bootstrapping will take time to download and install, the bootstrapper should have progress information and
+indicator during download and installation via stderr.
+Stderr is used instead of stdout so that user redirection of stdout only contains output from successful execution of their commandline.
+Use of stderr for informational messages is a convention used by native commands already.
+
+Should the install or download fail, the bootstrapper needs to output appropriate information and direct user to documentation on alternate
+install method and return a non-zero exit code.
+
+### Auto-updating
+
+Upon startup of the bootstrapper (following rules of [admin vs non-admin](#admin-vs-non-admins)) will detect if the installed
```

Comment:
Shouldn't this really be part of Windows/Microsoft Update? I dislike the idea of `pwsh` trying to do even more work at startup and potentially change the ground that I'm working on by automatically updating. I feel like the bootstrap script if this goes ahead should just install the initial version and updates are managed through WU.

Reply:
I think the original reason not to just use WU was because it wasn't supported on Nanoserver, but since we've decided Nano is out of scope, we should revisit this

Reply:
As of today, I think WU will work only if the installation is done through MSI and has the "use windows update" checked. How would it work for non-admin scenarios? I believe there will be many users using the non-admin installation.

Reply:
@jborean93 Agreed. I'm not comfortable with the idea that my set of build nodes will update to a newer version of PowerShell merely by running pwsh as part of a build.  I prefer to be "in control" of when that happens.

Reply:
Use winget as fallback.

Reply:
winget is not available on Windows Server and Windows Server support is required

Reply:
@SteveL-MSFT Since you say that Nano Server is out of scope, does that mean that PowerShell could be delivered via MSI after all (rather than Squirrel mentioned above)? IIRC, all SKUs of Windows other than Nano Server support Windows Installer.